User guide for Kubernetes Events in Ray Dashboard (#64734)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/k8s-events.md

> (kuberay-k8s-events)=

Reviewer: ```suggestion
---
myst:
  html_meta:
    description: "How to enable platform events in the Ray Dashboard on Kubernetes: RBAC permissions, the environment variable that turns the feature on, and how to verify that the dashboard receives events."
---

(kuberay-k8s-events)=
```

The style guide requires a `description` in the MyST `html_meta` front matter on every page, because search engines and the social preview use it. See the Front matter section of [`doc/source/ray-contribute/writing-style.md`](https://github.com/ray-project/ray/blob/master/doc/source/ray-contribute/writing-style.md).

## Overview
The Platform Events feature allows you to see Kubernetes cluster-level lifecycle events for Ray custom resources (RayCluster, RayJob, RayService) alongside your Ray application metrics and logs. This helps you correlate cluster behavior with application behavior.

## Prerequisites
* **Ray Version**: v2.56.0 or newer
* **Kubernetes Python Client**: The `kubernetes` Python package must be installed in the Ray Head Pod's Python environment. This is included in the official `rayproject/ray` images. If using custom images, ensure this package is added.
* **Kubernetes Cluster**: A functional Kubernetes cluster where you can deploy Ray workloads.

## Configure RBAC Permissions
The Ray Head Pod needs permissions to interact with the Kubernetes API to watch for relevant events.

Create a file named `ray-rbac.yaml`:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: ray-head-watch-role
rules:
- apiGroups: [""]
  resources: ["events"]
  verbs: ["get", "list", "watch"]
- apiGroups: ["ray.io"]
  resources: ["rayclusters"]

>   verbs: ["get"]

Reviewer: **Technical, for KubeRay SME review.** I think this Role is missing a rule for `pods`.

On `master`, the Kubernetes provider maintains a pod-membership cache and watches pod events. [`_relist_pod_membership`](https://github.com/ray-project/ray/blob/master/python/ray/dashboard/modules/platform_events/providers/k8s_provider.py) calls `list_namespaced_pod` with a `ray.io/cluster=<name>` label selector, and `_run_pod_membership_watch` and `_run_pod_events_watch` run as separate threads off the back of it. With the Role as written, the ServiceAccount can't list or watch pods, so that discovery fails.

That code landed in #63937 (merged 2026-07-22) and isn't in the `ray-2.56.0` or `ray-2.56.1` trees, so whether it belongs in this guide depends on what version you're documenting for. See my note on the Future enhancements section.

If you do add a `pods` rule, the "two permissions" count in my suggestion below needs to become three.

Author: > That code landed in #63937 (merged 2026-07-22) and isn't in the ray-2.56.0 or ray-2.56.1 trees, so whether it belongs in this guide depends on what version you're documenting for

exactly, since the guide is for 2.56.0, Im not sure if we want to include the rolebinding for Pods.

---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: ray-head-watch-role-binding
subjects:
- kind: ServiceAccount

>   name: default # IMPORTANT: Change this if your Ray Head Pod uses a different ServiceAccount

Reviewer: **Technical, for KubeRay SME review.** RBAC validation requires a `namespace` on a `ServiceAccount` subject, even in a namespaced `RoleBinding`. As written I'd expect `kubectl apply` to reject this with `subjects[0].namespace: Required value`. Worth applying the manifest verbatim to confirm.

Author: Added `namespace: default` under RoleBinding.subject for ServiceAccount

roleRef:
  kind: Role
  name: ray-head-watch-role
  apiGroup: rbac.authorization.k8s.io
```

Apply this configuration to the namespace where your Ray cluster runs:
```bash
kubectl apply -f ray-rbac.yaml -n <your-ray-namespace>
```

**Key Permissions:**
* `events`: Required to stream Kubernetes events.
* `rayclusters`: Required for the dashboard to find parent resources like RayJob or RayService and watch their related events.

## Configure Environment Variables
Enable and configure the feature using environment variables on the Ray Head Pod container.

| Environment Variable | Value | Description |
|----------------------|-------|-------------|

> | `RAY_DASHBOARD_INGEST_PLATFORM_EVENTS` | `"true"` | Enables the platform event watcher in the Ray Dashboard backend. |

Reviewer: ```suggestion
## Enable platform events

Set `RAY_DASHBOARD_INGEST_PLATFORM_EVENTS` to `"true"` on the Ray head pod container. This environment variable starts the platform event watcher in the Ray Dashboard backend.
```

A one-row table costs the reader a format switch for a single fact, so this makes it prose. The style guide's Choosing a format section reserves tables for mapping topics to pages or comparing options side by side. If more variables are coming, keep the table and I'll withdraw this.

Also sentence case in the heading, and "with" rather than "using" in the instrumental sense.

**Technical, for KubeRay SME review.** Two more environment variables gate this feature, and neither is documented here. [`_init_k8s_client`](https://github.com/ray-project/ray/blob/master/python/ray/dashboard/modules/platform_events/providers/k8s_provider.py) returns `False` and logs a warning if either `RAY_CLUSTER_NAME` or `RAY_CLUSTER_NAMESPACE` is unset, so the feature silently does nothing. The code comments say the KubeRay operator sets both on every Ray cluster pod, which is exactly the kind of thing worth stating explicitly: a reader debugging an empty Platform Events tab has no other clue where to look. You already worked through this in the thread with @sampan-s-nayak on the non-KubeRay case, so I may be re-treading ground you've decided against.

Author: Removed the table and applied your suggestion, thanks.

> Technical, for KubeRay SME review. 

yep I see @andrewsykim 's comment [here](https://github.com/ray-project/ray/pull/64734#discussion_r3676978581) stating most docs assume KubeRay is used, so not sure whether to explicitly call out `RAY_CLUSTER_NAME` and `RAY_CLUSTER_NAMESPACE` as env vars since they get handled/injected by kuberay automatically

## Manifest Examples

You can deploy complete working examples directly from the KubeRay repository:

### RayCluster Example
```bash

> kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-cluster.platform-events.yaml

Reviewer: **Blocking, though you already know.** Both sample URLs 404 as of today:

```
ray-cluster.platform-events.yaml   404
ray-job.platform-events.yaml       404
```

Confirming what you told @andrewsykim: this needs ray-project/kuberay#5001 to merge first. Flagging it so it doesn't get lost, since a merged guide with two dead `kubectl apply` commands is worse than no guide.

```

### RayJob Example
```bash
kubectl apply -f https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/ray-job.platform-events.yaml
```

## Verification
After deploying your Ray Cluster/Job with the above configurations:

1. **Port-Forward the Dashboard:**
```bash
# Find your Ray head pod
kubectl get pods -n <your-ray-namespace> | grep head

# Port-forward
kubectl port-forward <HEAD_POD_NAME> -n <your-ray-namespace> 8265:8265
```

2. **Query the API (Optional):**
Test the endpoint to see if events are being collected:
```bash
curl http://localhost:8265/api/v0/platform_events
```

A successful response will be a JSON object containing recent events:
```json
{
  "result": true,
  "msg": "Retrieved platform events",
  "data": {

>     "events": [

Reviewer: Question rather than a finding: do all four events need to be here? Reading them side by side, three are `raycluster-controller` `CreatedService`/`CreatedHeadPod`/`CreatedWorkerPod` events that differ only in `reason`, `objectName`, and `message`, and the fourth is the `rayjob-controller` `CreatedRayCluster` event. Two events, one from each controller, would show the same shape and the same parent-resource linking in about half the vertical space, and readers scanning for the response structure would get there faster. Your call, since you know what people ask about this payload.

Author: Do they *need* to be there? Perhaps no. I just included them to give examples of the different types of events we capture thats all. If the payload size is bothersome, I can trim it down.

## Troubleshooting & Important Notes

* **Requires KubeRay (Ray 2.56)**: In Ray 2.56, this feature only tracks events related to KubeRay Custom Resources (`RayCluster`, `RayJob`, `RayService`). It is not supported for manual Ray deployments without the KubeRay Operator.
* **Missing RBAC Permissions**: If the ServiceAccount lacks `get` permissions on `rayclusters` within the `ray.io` API group, the system cannot link events to parent RayJob or RayService resources. This will result in only RayCluster-direct events being shown.
* **Event Export Routing (Optional)**: By default, platform events are sent to Ray's observability system using the event type `PLATFORM_EVENT`.
  * If you use a custom event export address via `RAY_DASHBOARD_AGGREGATOR_AGENT_EVENTS_EXPORT_ADDR`, you can forward these platform events to external logging/monitoring systems (see {ref}`Ray Event Export <ray-event-export>`).
  * If you filter event types using `RAY_ENABLE_PYTHON_RAY_EVENT_TYPES`, make sure to include `"PLATFORM_EVENT"` in the list (e.g., `RAY_ENABLE_PYTHON_RAY_EVENT_TYPES="PLATFORM_EVENT,SYSTEM_EVENT"`). Otherwise, platform events will not be exported.

## Future Enhancements

> The current implementation in Ray 2.56 focuses on events targeting Ray custom resources. In a future release (Ray 2.57+), coverage will expand to include generic Kubernetes pod-level events (e.g., OOMs, scheduling issues). At that time, this feature will become useful for non-KubeRay deployments as well.

Reviewer: ```suggestion

```

**Recommend cutting this section.** Two reasons.

The first is policy: a user guide isn't the right place to promise unreleased functionality. Roadmap statements age badly and readers can't act on them. The design doc or a release note is a better home.

The second is that this one is already out of date. Pod-level events aren't a future release; they're on `master` now. #63937 ("Capture K8s Pod events in Ray Dashboard Head") merged 2026-07-22, and `k8s_provider.py` on `master` has `_run_pod_membership_watch`, `_run_pod_events_watch`, and `_relist_pod_membership` calling `list_namespaced_pod` with a `ray.io/cluster` label selector. None of that is in the `ray-2.56.0` or `ray-2.56.1` trees. So as of this guide's merge, the sentence "In a future release (Ray 2.57+), coverage will expand to include generic Kubernetes pod-level events" describes code that's already landed.

That leaves you a decision I'd defer to you and @andrewsykim: document the feature as of 2.56 and add pod events when 2.57 ships, or document `master` and cover pod events now. Either is defensible, but the "As of Ray 2.56" scoping in my Troubleshooting suggestion and the missing `pods` RBAC rule both hang off that call. `master` docs render at `/en/master`, which tracks `master` rather than the last release, so a reader on that page today gets a guide that undersells what their build does.

The style guide also flags the specific words here under Avoid time-relative words: "The current implementation," "In a future release," "At that time."

Author: Right, but  [#63937 ](https://github.com/ray-project/ray/pull/63937) isnt included in a release yet. I am happy to document pod events by just documenting master, I just thought users would want to know what release version they can officially use for a feature since users dont really run master in production. 

@andrewsykim your call.

Author: Just found out that 2.58 (which will include #63937) will be cut on Monday, so I updated the user-guide to have a `For deployments without KubeRay (Ray 2.58+)**` section describing how non-kuberay deployments can still use the feature.

And also removed the ` Future Enhancements` section. Lmk if this works.

Reviewer 2: LGTM to reference 2.58, I don't think we need to mention "non-kuberay" deployments though, we generally assume KubeRay is used for simplicity but advanced users can usually figure out how to enable without kuberay
